src/cyclebane/node_values.py

```diff
@@ -0,0 +1,302 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2024 Scipp contributors (https://github.com/scipp)
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections import abc
+from typing import TYPE_CHECKING, Any, Hashable, Iterable, Mapping, Sequence
+
+if TYPE_CHECKING:
+    import numpy
+    import pandas
+    import scipp
+    import xarray
+
+IndexName = Hashable
+IndexValue = Hashable
+
+
+class ValueArray(ABC):
```

Comment:
Good idea. But not easily extensible for users because the concrete type is selected within `NodeValues._to_value_arrays`. If this is a concern, I think you should add a mechanism to allow users to register their own `ValueArrays`. E.g.
```python
_VALUE_ARRAY_IMPLEMENTATIONS: dict[type, type] = {}
try:
    import pandas as pd
    _VALUE_ARRAY_IMPLEMENTATIONS[pd.Series] = PandasSeriesAdapter
except ModuleNotFoundError:
    pass

# ...

def register_value_array_adapter(key: type, adapter: type) -> None:
    _VALUE_ARRAY_IMPLEMENTATIONS[key] = adapter
```
And then select an implementation based on the type. (This may need a fallback to `SequenceAdapter` if the type is not in the map because there is no unique key for sequences.)

Reply:
Not a concern for now, I'd think? Supporting Pandas, Numpy, list, Xarray, and Scipp should get us pretty far?

Reply:
Added a registry implementation (different from suggestion) after all, while addressing the other comment in instance checks.